File: core/src/com/inkostilation/pong/engine/IEngine.java
package com.inkostilation.pong.engine;

import com.inkostilation.pong.commands.AbstractResponseCommand;
import com.inkostilation.pong.commands.AbstractRequestCommand;
import com.inkostilation.pong.exceptions.NoEngineException;

import java.io.IOException;

public interface IEngine<M> {

    void act(float delta);
    void receiveCommand(AbstractResponseCommand command, M mark) throws IOException, NoEngineException;
    void sendCommand(AbstractRequestCommand<IEngine<M>, M>... command) throws IOException, NoEngineException;
    void quit(M marker) throws IOException;
}

File: core/src/com/inkostilation/pong/server/engine/IEngine.kt
package com.inkostilation.pong.server.engine

import com.inkostilation.pong.commands.AbstractRequestCommand
import com.inkostilation.pong.commands.AbstractResponseCommand
import com.inkostilation.pong.server.network.Redirect
import java.util.*

interface IEngine {

    fun start()
    fun act(delta: Float)

    fun enter(marker: UUID)
    fun process(command: AbstractRequestCommand<AbstractEngine>): Array<AbstractResponseCommand<*>>
    fun quit(marker: UUID)

    fun getRedirect(): Redirect

    fun stop()
}